### PR TITLE
MOTECH-2559 Commcare tasks, import forms action

### PR DESCRIPTION
**Pull request by lkukielka**

### CODE REVIEW

File: commcare/src/main/java/org/motechproject/commcare/events/constants/DisplayNames.java [hunk 1]
```diff
@@ -30,6 +30,8 @@
     public static final String TYPE = "commcare.type";
     public static final String EXTRA_DATA = "commcare.extraData";
 
+    public static final String IMPORT_FORMS = "Import Forms";
```
Comment (sebbrudzinski) [inline commcare/src/main/java/org/motechproject/commcare/events/constants/DisplayNames.java]: Place it where the other task action names are (Create Case, Submit Form...)

File: commcare/src/main/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandler.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+@Component
+public class ImportFormActionEventHandler {
+
+    @Autowired
+    private EventRelay eventRelay;
+
+    @Autowired
+    private CommcareFormService formService;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ImportFormActionEventHandler.class);
```
Comment (sebbrudzinski) [inline commcare/src/main/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandler.java]: constants first

File: commcare/src/main/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandler.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    @MotechListener(subjects = EventSubjects.IMPORT_FORMS + ".*")
+    public void handleEvent(MotechEvent event) {
+
+        String configName = EventSubjects.getConfigName(event.getSubject());
+        DateTime startDate = (DateTime) event.getParameters().get(START_DATE);
+        DateTime endDate = (DateTime) event.getParameters().get(END_DATE);
+        Range<DateTime> dateRange = new Range<>(startDate, endDate);
+
+        CommcareFormImporterImpl importer = new CommcareFormImporterImpl(eventRelay, formService);
```
Comment (sebbrudzinski) [inline commcare/src/main/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandler.java]: We should make this a class field, so only one importer is used for all task-triggered imports

File: commcare/src/main/java/org/motechproject/commcare/pull/CommcareFormImporterImpl.java [hunk 4]
```diff
@@ -194,6 +197,8 @@ private void importFormList(CommcareFormList formList) {
         for (CommcareForm form : Lists.reverse(formList.getObjects())) {
             FullFormEvent formEvent = new FullFormEvent(form.getForm(), form.getReceivedOn(), form.getConfigName());
 
+            lastStartedImportingFormId = form.getId();
```
Comment (sebbrudzinski) [inline commcare/src/main/java/org/motechproject/commcare/pull/CommcareFormImporterImpl.java]: Why do we need this? How is this different from "lastImportedFormId"?

File: commcare/src/main/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandler.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import org.springframework.stereotype.Component;
+
+import static org.motechproject.commcare.events.constants.EventDataKeys.END_DATE;
+import static org.motechproject.commcare.events.constants.EventDataKeys.START_DATE;
+
+/**
+ * Class responsible for handling "Import Forms" action in tasks.
+ */
+@Component
+public class ImportFormActionEventHandler {
```
Comment (sebbrudzinski) [inline commcare/src/main/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandler.java:22]: Needs test

File: commcare/src/main/java/org/motechproject/commcare/tasks/builder/ImportFormActionBuilder.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import static org.motechproject.tasks.domain.mds.ParameterType.DATE;
+
+/**
+ * Responsible for building actions for importing forms for all configurations. There action can then be used in
+ * the task module.
+ */
+public class ImportFormActionBuilder implements ActionBuilder {
```
Comment (sebbrudzinski) [inline commcare/src/main/java/org/motechproject/commcare/tasks/builder/ImportFormActionBuilder.java:24]: Needs test

File: commcare/src/main/java/org/motechproject/commcare/pull/CommcareFormImporterImpl.java (hunk 4 above)
Comment (lkukielka) [inline commcare/src/main/java/org/motechproject/commcare/pull/CommcareFormImporterImpl.java]: _lastImportedFormId_ contains last successfully imported form id, but on error we need to log which form was the last to be imported

Comment (sebbrudzinski) [inline commcare/src/main/java/org/motechproject/commcare/pull/CommcareFormImporterImpl.java]: But doesn't an error indicate that the form was not actually imported?

**Comment by lkukielka** [comment]
test this please

File: commcare/src/main/java/org/motechproject/commcare/pull/CommcareFormImporterImpl.java [hunk 7]
```diff
@@ -111,11 +112,13 @@ public void run() {
                         }
                     } catch (RuntimeException e) {
                         LOGGER.error("Error while importing forms", e);
+                        LOGGER.error("{} of {} forms imported.", importCount, totalCount);
                         handleImportError(e, configName);
                     }
                 } while (importInProgress && hasMore);
 
-                LOGGER.info("Form import finished");
+                LOGGER.info("Form import finished. {} of {} forms imported. Last form xmlns to be imported was {}", importCount, totalCount, lastFormXMLNSToBeImported);
```
Comment (sebbrudzinski) [inline commcare/src/main/java/org/motechproject/commcare/pull/CommcareFormImporterImpl.java]: The Last form XMLNS to be imported was meant to be logged when there's an error, right? Why was this changed?

File: commcare/src/test/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    public void shouldCallImporterMethodsWithCorrectArguments() {
+        Range<DateTime> range = new Range<>(DATES.get(0), DATES.get(1));
+
+        importer = Mockito.mock(CommcareFormImporterImpl.class);
+        when(importer.countForImport(eq(range), eq(CONFIG_NAME))).thenReturn(0);
+        doNothing().when(importer).startImport(eq(range), eq(CONFIG_NAME));
+        eventHandler.setImporter(importer);
+
+        MotechEvent event = prepareEvent(true);
+        eventHandler.handleEvent(event);
```
Comment (sebbrudzinski) [inline commcare/src/test/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandlerTest.java:68]: This tests nothing

File: commcare/src/test/java/org/motechproject/commcare/tasks/builder/ImportFormActionBuilderTest.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+        when(configService.getConfigs()).thenReturn(configs);
+
+        importFormActionBuilder = new ImportFormActionBuilder(configService);
+    }
+
+    @Test
+    public void buildActionsShouldReturnTwoActionsForTwoConfigs() {
+        List<ActionEventRequest> eventRequests = importFormActionBuilder.buildActions();
+        assertEquals(2, eventRequests.size());
```
Comment (sebbrudzinski) [inline commcare/src/test/java/org/motechproject/commcare/tasks/builder/ImportFormActionBuilderTest.java:40]: Expand this a little, verify event request fields

File: commcare/src/test/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandlerTest.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    public void setUp() {
+        initMocks(this);
+        eventHandler = new ImportFormActionEventHandler();
+    }
+
+    @Test
+    public void shouldCallImporterMethodsWithCorrectArguments() {
+        Range<DateTime> range = new Range<>(DATES.get(0), DATES.get(1));
+
+        importer = Mockito.mock(CommcareFormImporterImpl.class);
```
Comment (sebbrudzinski) [inline commcare/src/test/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandlerTest.java]: Why don't we use annotation for this mock as well?

File: commcare/src/main/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandler.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    private EventRelay eventRelay;
+
+    @Autowired
+    private CommcareFormService formService;
+
+    private CommcareFormImporterImpl importer;
+
+    @PostConstruct
+    public void init() {
+        importer = new CommcareFormImporterImpl(eventRelay, formService);
```
Comment (sebbrudzinski) [inline commcare/src/main/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandler.java]: We will need to create it per event. Let's create a separate factory that will return instances of the importer.

File: commcare/src/main/java/org/motechproject/commcare/pull/CommcareTasksFormImporterFactory.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+@Component
+public class CommcareTasksFormImporterFactory {
+
+    @Autowired
+    private EventRelay eventRelay;
+
+    @Autowired
+    private CommcareFormService formService;
+
+    private final Map<String, CommcareFormImporterImpl> importerMap = new HashMap<>();
```
Comment (sebbrudzinski) [inline commcare/src/main/java/org/motechproject/commcare/pull/CommcareTasksFormImporterFactory.java]: We do not need that map. Just make the method return a new instance of CommcareFormImporterImpl

File: commcare/src/main/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandler.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        Range<DateTime> dateRange = new Range<>(startDate, endDate);
+
+        int formsToImport = importer.countForImport(dateRange, configName);
+
+        LOGGER.info("{} commcare forms to be imported.", formsToImport);
+
+        importer.startImport(dateRange, configName);
+    }
+
+    public void setImporterFactory(CommcareTasksFormImporterFactory importerFactory) {
```
Comment (sebbrudzinski) [inline commcare/src/main/java/org/motechproject/commcare/events/imports/ImportFormActionEventHandler.java]: This shouldn't have a setter

**Comment by sebbrudzinski** [comment]
@pgesek LGTM
